File: tablet/Resources/task/QoLSG.js
// QoLSG.js

/*
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    tablename = "qolsg",
    fieldlist = dbcommon.standardTaskFields(),
    EVENT_FROM_WEBVIEW = 'QOLSG_EVENT_FROM_WEBVIEW',
    EVENT_TO_WEBVIEW = 'QOLSG_EVENT_TO_WEBVIEW';

fieldlist.push(
    // Results
    {name: 'category_start_time', type: DBCONSTANTS.TYPE_DATETIME},
    {name: 'category_responded', type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: 'category_response_time', type: DBCONSTANTS.TYPE_DATETIME},
    {name: 'category_chosen', type: DBCONSTANTS.TYPE_TEXT},
    {name: 'gamble_fixed_option', type: DBCONSTANTS.TYPE_TEXT},
    {name: 'gamble_lottery_option_p', type: DBCONSTANTS.TYPE_TEXT},
    {name: 'gamble_lottery_option_q', type: DBCONSTANTS.TYPE_TEXT},
    {name: 'gamble_lottery_on_left', type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: 'gamble_starting_p', type: DBCONSTANTS.TYPE_REAL},
    {name: 'gamble_start_time', type: DBCONSTANTS.TYPE_DATETIME},
    {name: 'gamble_responded', type: DBCONSTANTS.TYPE_BOOLEAN},
    {name: 'gamble_response_time', type: DBCONSTANTS.TYPE_DATETIME},
    {name: 'gamble_p', type: DBCONSTANTS.TYPE_REAL},
    {name: 'utility', type: DBCONSTANTS.TYPE_REAL}
);

// CREATE THE TABLES

dbcommon.createTable(tablename, fieldlist);

//=============================================================================
// TASK (inc. webview handler)
//=============================================================================

function QoLSG(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
}

lang.inheritPrototype(QoLSG, taskcommon.BaseTask);
lang.extendPrototype(QoLSG, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: QoLSG,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    _editable: false,

    // OTHER

    // Standard task functions
    isComplete: function () {
        return this.utility !== null;
    },

    getSummary: function () {
        return L("qolsg_utility") + " " + lang.toFixedOrNull(this.utility, 3) +
                this.isCompleteSuffix();
    },

    getDetail: function () {
        var msg = "",
            i;
        for (i = 0; i < fieldlist.length; ++i) {
            msg += (
                fieldlist[i].name + ": " +
                dbcommon.userString(this, fieldlist[i]) + "\n"
            );
        }
        return msg;
    },

    edit: function () {  // readOnly ignored
        var self = this,
            UICONSTANTS = require('common/UICONSTANTS'),
            moment = require('lib/moment'),
            params = {
                TEXT_INITIAL_INSTRUCTION: L("qolsg_initial_instruction"),
                TEXT_CURRENT_STATE: L("qolsg_current_state"),
                TEXT_DEAD: L("qolsg_dead"),
                TEXT_HEALTHY: L("qolsg_healthy"),
                TEXT_INDIFFERENT: L("qolsg_indifferent"),
                TEXT_BACK: L("back"),
                TEXT_THANKS: L("webview_thanks"),
                TEXT_H_ABOVE_1: L("qolsg_h_above_1"),
                TEXT_H_0_TO_1: L("qolsg_h_0_to_1"),
                TEXT_H_BELOW_0: L("qolsg_h_below_0"),
                TEXT_LEFT: L("qolsg_left"),
                TEXT_RIGHT: L("qolsg_right"),
                TEXT_INSTRUCTION_PREFIX_2: L("qolsg_instruction_prefix_2"),
                TEXT_INSTRUCTION_MEDIUM: L("qolsg_instruction_medium"),
                TEXT_INSTRUCTION_LOW: L("qolsg_instruction_low"),
                TEXT_INSTRUCTION_HIGH: L("qolsg_instruction_high"),
                TEXT_INSTRUCTION_SUFFIX: L("qolsg_instruction_suffix")
            },
            html = taskcommon.loadHtmlSetParams(
                "task_html/qolsg_task.html",
                params,
                "task_html/qolsg_task.jsx"
            ),
            editing_time_start,
            editing_time_end,
            editing_time_s,
            window;

        self.dbstore(); // sets self.id

        //---------------------------------------------------------------------
        // Communication with webview
        //---------------------------------------------------------------------

        function eventFromWebView(e) {
            Titanium.API.trace("QoLSG: eventFromWebView: " + e.eventType);
            var conversion = require('lib/conversion'),
                dict;
            switch (e.eventType) {

            case "savetrial":
                dict = JSON.parse(e.data,
                                  conversion.json_decoder_reviver);
                lang.copyProperties(dict, self);
                self.dbstore();
                break;

            case "exit":
            case "abort":
                editing_time_end = moment();
                editing_time_s = editing_time_end.diff(editing_time_start,
                                                       'seconds',
                                                       true);  // floating-point
                self.defaultFinishedFn(
                    e.eventType === "exit" ?
                            UICONSTANTS.FINISHED :
                            UICONSTANTS.ABORTED,
                    editing_time_s
                );
                if (window !== null) {
                    window.close();
                    window = null;
                }
                break;
            }
        }

        //---------------------------------------------------------------------
        // Launch webview
        //---------------------------------------------------------------------
        window = taskcommon.createFullscreenWebviewWindow(
            html,
            EVENT_FROM_WEBVIEW,
            eventFromWebView
        );
        editing_time_start = moment();
        window.open();
    }

});

module.exports = QoLSG;

File: server/tasks/qolsg.py
#!/usr/bin/env python3
# qolsg.py

"""
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

import cardinal_pythonlib.rnc_web as ws
from cc_modules.cc_constants import (
    CTV_DICTLIST_INCOMPLETE,
    PV,
)
from cc_modules.cc_html import (
    get_yes_no_none,
    tr_qa,
)
from cc_modules.cc_task import Task


# =============================================================================
# QoL-SG
# =============================================================================

DP = 3


class QolSG(Task):
    tablename = "qolsg"
    shortname = "QoL-SG"
    longname = "Quality of Life: Standard Gamble"
    fieldspecs = [
        dict(name="category_start_time", cctype="TEXT",
             comment="Time categories were offered (ISO-8601)"),
        dict(name="category_responded", cctype="INT", pv=PV.BIT,
             comment="Responded to category choice? (0 no, 1 yes)"),
        dict(name="category_response_time", cctype="TEXT",
             comment="Time category was chosen (ISO-8601)"),
        dict(name="category_chosen", cctype="TEXT",
             comment="Category chosen: high (QoL > 1), "
             "medium (0 <= QoL <= 1), low (QoL < 0)"),
        dict(name="gamble_fixed_option", cctype="TEXT",
             comment="Fixed option in gamble (current, healthy, dead)"),
        dict(name="gamble_lottery_option_p", cctype="TEXT",
             comment="Gamble: option corresponding to p  "
             "(current, healthy, dead)"),
        dict(name="gamble_lottery_option_q", cctype="TEXT",
             comment="Gamble: option corresponding to q  "
             "(current, healthy, dead) (q = 1 - p)"),
        dict(name="gamble_lottery_on_left", cctype="INT", pv=PV.BIT,
             comment="Gamble: lottery shown on the left (0 no, 1 yes)"),
        dict(name="gamble_starting_p", cctype="FLOAT", min=0, max=1,
             comment="Gamble: starting value of p"),
        dict(name="gamble_start_time", cctype="TEXT",
             comment="Time gamble was offered (ISO-8601)"),
        dict(name="gamble_responded", cctype="INT", pv=PV.BIT,
             comment="Gamble was responded to? (0 no, 1 yes)"),
        dict(name="gamble_response_time", cctype="TEXT",
             comment="Time subject responded to gamble (ISO-8601)"),
        dict(name="gamble_p", cctype="FLOAT", min=0, max=1,
             comment="Final value of p"),
        dict(name="utility", cctype="FLOAT",
             comment="Calculated utility, h"),
    ]

    def get_trackers(self):
        return [
            {
                "value": self.utility,
                "plot_label": "Quality of life: standard gamble",
                "axis_label": "QoL (0-1)",
                "axis_min": 0,
                "axis_max": 1,
            }
        ]

    def get_clinical_text(self):
        if not self.is_complete():
            return CTV_DICTLIST_INCOMPLETE
        return [{
            "content":  "Quality of life: {}".format(
                        ws.number_to_dp(self.utility, DP))
        }]

    def is_complete(self):
        return (
            self.utility is not None and
            self.field_contents_valid()
        )

    def get_task_html(self):
        h = """
            <div class="summary">
                <table class="summary">
        """ + self.get_is_complete_tr()
        h += tr_qa("Utility", ws.number_to_dp(self.utility, DP, default=None))
        h += """
                </table>
            </div>
            <div class="explanation">
                Quality of life (QoL) has anchor values of 0 (none) and 1
                (perfect health). The Standard Gamble offers a trade-off to
                determine utility (QoL).
                Values &lt;0 and &gt;1 are possible with some gambles.
            </div>
            <table class="taskdetail">
                <tr><th width="50%">Measure</th><th width="50%">Value</th></tr>
        """
        h += tr_qa("Category choice: start time", self.category_start_time)
        h += tr_qa("Category choice: responded?",
                   get_yes_no_none(self.category_responded))
        h += tr_qa("Category choice: response time",
                   self.category_response_time)
        h += tr_qa("Category choice: category chosen", self.category_chosen)
        h += tr_qa("Gamble: fixed option", self.gamble_fixed_option)
        h += tr_qa("Gamble: lottery option for <i>p</i>",
                   self.gamble_lottery_option_p)
        h += tr_qa("Gamble: lottery option for <i>q</i> = 1 – <i>p</i>",
                   self.gamble_lottery_option_q)
        h += tr_qa("Gamble: lottery on left?",
                   get_yes_no_none(self.gamble_lottery_on_left))
        h += tr_qa("Gamble: starting <i>p</i>", self.gamble_starting_p)
        h += tr_qa("Gamble: start time", self.gamble_start_time)
        h += tr_qa("Gamble: responded?",
                   get_yes_no_none(self.gamble_responded))
        h += tr_qa("Gamble: response time", self.gamble_response_time)
        h += tr_qa("Gamble: <i>p</i>",
                   ws.number_to_dp(self.gamble_p, DP, default=None))
        h += tr_qa("Calculated utility",
                   ws.number_to_dp(self.utility, DP, default=None))
        h += """
            </table>
        """
        return h
